### ISSUE TITLE
WorkloadInstanceHandler and queueing is broken

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#45114]
Example test failure: https://prow.istio.io/view/gs/istio-prow/logs/unit-tests_istio_postsubmit/1661434824121913344

Problem:

SE and kube controller have queues. Their readiness is based on this queue. All events need to be serialized to avoid logical races.

WorkloadInstanceHandler does not use a queue, which leads to a breakage (in the above test, we build 1 endpoint then set EDS to that, reverting a concurrent change to (correctly) set it to 2 endpoints).

However, just putting a queue in WorkloadInstanceHandler is also broken. We need all initially known workloads to be handled before sync

Basically we have this case

```mermaid
graph TD
    subgraph K8s
        Event1-->Event2-->Ready-->WorkloadHandler
    end
    subgraph SE
        EventA-->EventB-->ready
        EventB-.->WorkloadHandler
    end
```

We will mark SE as "ready" while WorkloadHandler is processed async.

It seems like we need WorkloadHandler to be enqueued, but also block. This seems to work but seems pretty odd

**Comment by howardjohn** [comment]
Really the proper fix is to not use Handlers at all, but treat all things as streams/collections/informers of data. So we would have a WorkloadInstance informer. which leads back to https://docs.google.com/document/d/1-ywpCnOfubqg7WAXSPf4YgbaFDBEU9HIqMWcxZLhzwE/edit

**Comment by howardjohn** [comment]
technically we only need to block until startup is done...

**Comment by hzxuzhonghu** [comment]
That's interesting, we have run it for years without finding any issue

**Comment by howardjohn** [comment]
Yes, I am not sure why it broke recently. I checked it is NOT from #44145 which I thought it might be; it was broken before then.

It is a pretty rare flake, takes like 10k runs to reproduce on my machine typically.

**Comment by hzxuzhonghu** [comment]
OK, we can mark it as low priority and track it here
